drv/lpc55-sprot-server/src/main.rs

```diff
@@ -175,280 +114,618 @@ fn main() -> ! {
+        State::Read(ReadState::WaitingForRequest)
+    }
+
+    /// Return the read state. Panics if the RoT is not currently reading.
+    fn read(&self) -> ReadState {
+        match self {
+            State::Read(s) => *s,
+            _ => panic!(),
         }
     }
+
+    /// Return the write state. Panics if the RoT is not currently writing.
+    fn write(&self) -> WriteState {
+        match self {
+            State::Write(s) => *s,
+            _ => panic!(),
+        }
+    }
+}
+
+/// The state of the IO state machine when trying to read a request from the SP
+#[derive(Clone, Copy, PartialEq)]
+pub enum ReadState {
+    /// The RoT is waiting for CSn to be asserted
+    WaitingForRequest,
+
+    /// The RoT is in frame and reading data until it sees CSn de-asserted
+    InFrame,
+
+    /// The frame was read cleanly.
+    ///
+    /// In this case we were:
+    ///   1. In `WaitingForRequest` and saw CSn asserted and CSn de-asserted at the
+    ///      same time with no overrun errors, and with data.
+    ///   2. In `InFrame` and saw CSn de-asserted with no overrun errors and
+    ///     with data.
+    ///
+    FrameRead,
+
+    /// The SP has pulsed CSn, which means that CSn was asserted and de-
+    /// asserted with no data clocked at all.
+    ///
+    /// The SP pulses CSn when it receives a ROT_IRQ it wasn't expecting. This
+    /// happens in the reply phase, and can occur if processing a request at
+    /// the RoT takes longer than a retry timeout at the SP.
+    ///
+    /// We can only transition to Flush in the `WaitingForRequest` or `InFrame`
+    /// states. Flush is a `transient` state indicating we should clear our
+    /// FIFOs and buffers and go back to `WaitingForRequest`.
+    Flush,
+
+    /// We missed reading some bytes. We should clear our buffers and fifos and
+    /// inform the SP.
+    Overrun,
+
+    /// We got an unexpected CSn Assert. This is either due to a weird timing
+    /// issue resulting in the next request being sent, or a CSn pulse.
+    ///
+    /// Either way we are desynchronized, and the safest thing to do is
+    /// to clear our buffers and fifos and go back to `WaitingForRequest`
+    /// *without* sending a reply. We do *NOT* want to send a reply for the
+    /// following three reasons:
+    ///
+    /// 1. The CSn assert is the start of a CSn pulse. Sending a reply triggers
+    /// an ROT_IRQ assert which is exactly what the CSn pulse is trying to get
+    /// rid of!
+    ///
+    /// 2. The CSn assert is the start of a new request and the new request
+    /// fits in one fifo. In this case the response from the prior request
+    /// will be interpreted as the response for the new request. While we
+    /// could use sequence numbers to clear up this ambiguity, doing so adds
+    /// complexity.
+    ///
+    /// 3. The CSn assert is the start of a new request and the new request
+    /// does not fit in one fifo. We will go process the prior response, and
```

Comment:
Again, be specific, that we don't want to handle a request received the first time through the loop and then come back and process the request from the new CSn assert. The current code doesn't do that.

Note to other reviewers, this comment was written to structure my thoughts before the code was written.